lib/blocks.php

```diff
@@ -0,0 +1,62 @@
+<?php
+/**
+ * Overrides all core blocks.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Registers block type from metadata file.
+ *
+ * @param string $file   Path to the translation file to load.
+ */
+function gutenberg_register_block_type_from_metadata( $file ) {
+	$metadata = json_decode( file_get_contents( $file ), true );
+	if ( empty( $metadata['name'] ) || empty( $metadata['title'] ) || empty( $metadata['category'] ) ) {
+		$message = __( 'Required block types must be provided.' );
+		_doing_it_wrong( __FUNCTION__, $message, '5.2.0' );
+		return;
+	}
+	$block_name = $metadata['name'];
+	$settings = array(
+		'title' => $metadata['title'],
+		'category' => $metadata['category'],
+	);
+	foreach( array( 'description', 'keywords', 'attributes', 'supports' ) as $field_name ) {
+		if ( ! empty( $metadata[ $field_name ] ) ) {
+			$settings[ $field_name ] = $metadata[ $field_name ];
+		}
+	}
+	if ( ! empty( $metadata['renderCallback'] ) ) {
+		$render_callback_file = dirname( $file ) . '/' . $metadata['renderCallback'];
+		if ( file_exists( $render_callback_file ) ) {
+			require $render_callback_file;
+			$settings['render_callback'] = 'gutenberg_render_block_' . str_replace( array( '/', '-' ), '_', $block_name );
```

Comment:
I think it requires [anonymous functions](http://php.net/manual/en/functions.anonymous.php) (PHP 5.4.0+), but I wonder if rather than relying on a specific naming convention for functions, we use the file returns semantics:

https://gist.github.com/aduth/ef27bfaae19a9959a263f1600b37c71d

Reply:
The `register_block_render_callback` could impose more requirements on the runtime which consumes from the file (i.e. the function needs to be defined somewhere), in case that might deter us.

A yet-other idea is treating it more like a template file, expecting it to output HTML directly. I'm not sure this is a good way to go though, and it's not obvious how the template would read from attributes (unless they were just some `$attributes` global defined for use by the file).

Reply:
Some of those existing render callbacks reference other custom functions so this might get tricky. I don't know, maybe this would work. I'd love to see some thoughts from developers better familiar with patterns used in other parts of WordPress core like @pento or @danielbachhuber.

Reply:
I'm not 100% sure I follow. Why is `$metadata['renderCallback']` a file and not simply a callable function?

Probably the best example of prior art to follow is `register_shortcode()` (which accepts a callable function). I wouldn't be opposed to taking the template file approach instead, but core doesn't currently have the pattern of injecting variables into a template (see the [long conversation in 21676](https://core.trac.wordpress.org/ticket/21676))